mlflow/recipes/steps/predict.py

```diff
@@ -198,6 +175,56 @@ def _run(self, output_directory):
             _PREDICTION_COLUMN_NAME, predict(struct(*input_sdf.columns))
         )
 
+        # check if output location is already populated for non-delta output formats
+        output_format = self.step_config["using"]
+        output_location = self.step_config["location"]
+        output_populated = False
+        if self.save_mode in ["default", "error", "errorifexists"]:
+            if output_format == "parquet" or output_format == "delta":
+                output_populated = os.path.exists(output_location)
+            else:
+                try:
+                    output_populated = spark._jsparkSession.catalog().tableExists(output_location)
+                except Exception:
+                    # swallow spark failures
+                    pass
+        if output_populated:
+            raise MlflowException(
+                message=(
+                    f"Output location `{output_location}` using format `{output_format}` is "
+                    "already populated. To overwrite, please change the spark `save_mode` in "
+                    "the predict step configuration."
+                ),
+                error_code=BAD_REQUEST,
+            )
+
+        if output_format == "table":
+            try:
+                from delta.tables import DeltaTable
+
+                output_populated = DeltaTable.forName(spark, output_location)
+            except Exception:
+                # swallow spark failures
+                pass
+
+            if output_populated:
+                _logger.info(f"Table already exists at {output_location}")
+                spark.sql(
+                    f"ALTER TABLE {output_location} SET TBLPROPERTIES ('delta.columnMapping.mode'='name','delta.minReaderVersion'='2','delta.minWriterVersion'='5')"
+                )
+            else:
+                _logger.info(f"Creating a new table at {output_location}")
+                from delta.tables import DeltaTable
+
+                DeltaTable.create().addColumns(scored_sdf.schema).property(
+                    "delta.minReaderVersion", "2"
+                ).property("delta.minWriterVersion", "5").property(
+                    "delta.columnMapping.mode", "name"
+                ).tableName(
+                    output_location
+                ).execute()
+                self.save_mode = "append"
```

Comment:
Why is this needed given it's a new table to be created in `else` branch? `save_mode` is user input. I don't think it's a good idea to overwrite it w/o good reason.

Reply:
So saving to the table always checks if the table is created. 
Previous the save was implicitly creating the table if the one was not created. Now we are creating a table and saving it post that. So if the table already exists, the default save_mode doesn't work, so we need to append it. 
We are changing the save_mode only in the create case and not in the case where the table already exists. Let me know if I am missing something

let me add comments on why we are doing this in the code so it is easier for the reader to understand.

Reply:
Got it. Makes sense to me now! Thanks!